daemon/metrics.go

```diff
@@ -40,3 +49,64 @@ func init() {
 	imageActions = ns.NewLabeledTimer("image_actions", "The number of seconds it takes to process each image action", "action")
 	metrics.Register(ns)
 }
+
+const metricsPluginType = "MetricsCollector"
+
+func (d *Daemon) cleanupMetricsPlugins() {
+	ls := d.PluginStore.GetAllManagedPluginsByCap(metricsPluginType)
+	var wg sync.WaitGroup
+	wg.Add(len(ls))
+
+	for _, p := range ls {
+		go func() {
+			defer wg.Done()
+			stopMetricsPlugin(p)
+		}()
+	}
+	wg.Wait()
+
+	if d.metricsPluginListener != nil {
+		d.metricsPluginListener.Close()
+	}
+}
+
+type metricsPlugin struct {
+	plugingetter.CompatPlugin
+}
+
+func (p metricsPlugin) sock() string {
+	return "metrics.sock"
+}
+
+func (p metricsPlugin) sockBase() string {
+	return filepath.Join(p.BasePath(), "run", "docker")
+}
+
+func startMetricsPlugin(p plugingetter.CompatPlugin) error {
```

Comment:
Its a bit confusing to call this `StartMetrics` and `StopMetrics`, when in reality its a signal from the daemon to the plugin that the metrics socket is available and unavailable. 

Also the errors are misleading. `could not start metrics plugin` is not accurate. The plugin is started irrespective of this API call using `docker plugin enable`.